### PR TITLE
[FLINK-15458][conf][docs] Add support for whitelisting ambiguous options 

### PR DESCRIPTION
**Pull request by zentol**
The `ConfigOptionsDocsCompletenessITCase` verifies that all existing and documented options are well-defined; as in that for any key there is exactly 1 default and description present, globally.

There is one use-case however where this check is too strict: reporters. These only work with key suffixes (like "port") and hence are failing this check. As a result we don't use the generator mechanism for reporters, which is unfortunate.

This PR introduces a whitelisting mechanism that allows us to define a set of keys for which this check is not performed.
Note that the remaining checks of the test are still applied. (Is the option documented, does the documented option exist, do the defaults/description match).

The changes are mostly a refactoring of the test. Previously the test was checking whether an ambiguous key exists during the discovery of keys, which was now separated.
This implies that during discovery we no longer end up with a 1:1 mapping of key:option, but rather 1:N which is later checked.

### CODE REVIEW

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java [hunk 1]
```diff
@@ -61,50 +64,103 @@
 
 	private static final Formatter htmlFormatter = new HtmlFormatter();
 
+	// options for which we allow distinct definitions
+	// this allows reporters to define their own options that are technically only key suffixes
+	private static final Set<String> WELL_DEFINED_WHITELIST = new HashSet<>(Arrays.asList(
```
Comment (Myasuka) [inline flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: I think we missed `host` and `port` here.

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java [hunk 2]
```diff
@@ -61,50 +64,103 @@
...
+				}
+			});
+	}
+
+	private static <O> void assertOptionsAreWellDefined(Map<String, List<O>> allOptions, BiFunction<O, O, AssertionError> duplicateHandler) {
+		allOptions.entrySet().stream()
+			.filter(entry -> !WELL_DEFINED_WHITELIST.contains(entry.getKey()))
+			.map(Map.Entry::getValue)
+			.forEach(options -> options.stream().reduce((option1, option2) -> {
+				if (option1.equals(option2)) {
```
Comment (Myasuka) [inline flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: I think we should better include `containingClass` to compare when calling `ExistingOption#equals` so that these two `ExistingOption`s are really identical.

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java (hunk 1 above)
Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: yes I did, happened while cleaning up the history...

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java (hunk 2 above)
Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: That would violate our definition of begin well-defined; from the javadocs:
`no 2 options exist for the same key with different descriptions/default values`

The organization into separate classes that we have is only for developer convenience and readability only; it does not (and must not) have any semantics attached to it. Hence the containing class doesn't matter.

The reason for this is simplicity; we don't have to worry about
* options clashing in their default value, which is difficult to document in a good way and can lead to subtle issues when de-duplicating options
* options clashing in their description, which is also difficult to document and usually leads to stale documentation at one place
* options clashing in their type, which may result in a component being unusable when used in conjunction with another
* users not being able to configure distinct values per option

Comment (Myasuka) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: If some developers happen to create the same config key in different classes, I‘m afraid there exists some risk that user configure that key in `flink-conf.yaml` but taken effect by another unexpected class.

By the way, after introducing data type for config option, we did not include that field to compare in [duplication check](https://github.com/apache/flink/blob/df5eb214294f09a3cd7fcc04ab2617a321041e9d/flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java#L219), and this is certainly out of scope of this PR.

Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: > If some developers happen to create the same config key in different classes, I‘m afraid there exists some risk that user configure that key in flink-conf.yaml but taken effect by another unexpected class.

That's exactly what this check is preventing.

Comment (Myasuka) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: Would you please kindly explain whether the scenario below is legal in Flink?  
We define two config option named as `a.b.c` with the same default value and description in different options files (e.g in `CheckpointingOptions` and `RocksDBOptions`). Current `ConfigOptionsDocsCompletenessITCase` would still pass successfully.

Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: That's legal. IIRC, when we introduced the generator we actually had an option that was mirroring another one. One could also argue that logically, if code-structure must not have semantics, duplication of code must be allowed.

However, it seems unlikely that 2 options are being created for which they
1. share a key
2. have an _exactly_ matching description (supposedly written in isolation from each other)
3. are semantically so aligned that the key/description are still meaningful
4. yet are semantically so different that the accidental configuration can cause problems.

Overall I think this is mostly a theoretical issue due to us usually providing a component context via the key; to take your example, all `RocksDBOptions` have `rocksdb` in their key, with the inverse rightfully being the case for the `CheckpointingOptions`.

Note that in any case the behavior in this regard isn't different to `master`.

Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: It is also not really the responsibility of this test to forbid this; which is apparent when you consider that whether you de-duplicate the option or not, the contents of the documentation remain the same.

Comment (Myasuka) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: Thanks for your interpretation, I still reserve my judgment on this problem and I agree this should be out of scope of this PR since this problem has been existed for a while.

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java [hunk 3]
```diff
@@ -61,50 +64,105 @@
...
+					// default is outdated
+					problems.add("Documented default of " + supposedState.key + " in " + supposedState.containingClass +
+						" is outdated. Expected: " + supposedState.defaultValue);
+				} else if (documentedState.stream().noneMatch(documentedOption -> supposedState.description.equals(documentedOption.description))) {
+					// description is outdated
+					problems.add("Documented description of " + supposedState.key + " in " + supposedState.containingClass +
+						" is outdated.");
+				} else {
+					// the docs for this option are up-to-date
+				}
```
Comment (tillrohrmann) [inline flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: I think we won't fail if we have a `DocumentedOption` which has no corresponding `ExistingOption` since we are only looking whether there is a `DocumentedOption` for every `ExistingOption`.

Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: That is checked just below this block by verifying that the set of documented options (effectively the set of _unmatched_ options) is empty.

Comment (zentol) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: However, something that could indeed slip through if there are say, 3 documented options with key X, but only 2 existing options with that key.

**Comment by zentol** [comment]
@tillrohrmann 
An annotation-based approach was my first instinct as well but that doesn't work with the existing checks the test makes.
At it's core, the test does 3 things, which with the structure is obvious:
```
1) assertDocumentedOptionsAreWellDefined(documentedOptions);
2) assertExistingOptionsAreWellDefined(existingOptions);
3) compareDocumentedAndExistingOptions(documentedOptions, existingOptions);
```

The check for documented options being well-defined is where I ran into issues; since annotations are only available for the actual ConfigOption there was no way to differentiate between normal and suffix options when you only look at the documentation.

Including the prefix in the documentation can solve it, but I don't want them in the documentation as they add a lot of noise and the reporter-specific configuration page becomes a lot less concise.

However, when you look at the core test structure, and think about it (as I just did trying to recall why I didn't go with annotations), then check 1) should be redundant. If all existing options are well-defined, and the documented and existing options match, then we can conclude that the documented options are also well-defined.
As such, if we were to remove this check then we could in fact go with an annotation approach.

I'll think about if a bit more.

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: Yes, this is what I meant.

**Comment by tillrohrmann** [comment]
Just another thought: Maybe one could add the prefix for suffix options as a non visible attribute to the key field of a configuration option row in the generated html output. Or we indeed say that we omit check `1)`.

In the latter case one would need to find the one matching option out of a collection of `DocumentedOption` with an `ExistingOption`.

**Comment by zentol** [comment]
I went with the latter approach of removing check `1)`; we have a logically sound argument that the check is redundant, and I prefer that over test-specific fields in the documentation.

Options and entire classes can now be annotated with `Documentation.SuffixOption`. Only if all options for a given key are suffix options they are excluded from the check for being well-defined.
